packages/react-ui/src/widgets/time-series-widget-ui/TimeSeriesWidgetUI.js

```diff
@@ -0,0 +1,424 @@
+import {
+  Box,
+  Menu,
+  Grid,
+  IconButton,
+  MenuItem,
+  SvgIcon,
+  Typography
+} from '@material-ui/core';
+import React, { useEffect, useMemo, useRef, useState } from 'react';
+import TimeSeriesChart from './components/TimeSeriesChart';
+import { TimeSeriesProvider, useTimeSeriesContext } from './hooks/TimeSeriesContext';
+import { CHART_TYPES } from './utils/constants';
+import { PropTypes } from 'prop-types';
+import { GroupDateTypes, getMonday } from '@carto/react-core';
+
+const FORMAT_DATE_BY_STEP_SIZE = {
+  [GroupDateTypes.DAYS]: daysCurrentDateRange,
+  [GroupDateTypes.WEEKS]: weeksCurrentDateRange
+};
+
+const FORMAT_TIMEFRAME_BY_STEP_SIZE = {
+  [GroupDateTypes.DAYS]: daysCurrentDateRange,
+  [GroupDateTypes.WEEKS]: daysCurrentDateRange
+};
+
+const TIMEFRAME_STEP_BY_STEP_SIZE = {
+  [GroupDateTypes.DAYS]: 60 * 60 * 24,
+  [GroupDateTypes.WEEKS]: 60 * 60 * 24
+};
+
+function TimeSeriesWidgetUI({
+  data,
+  stepSize,
+  chartType,
+  duration,
+  tooltip,
+  tooltipFormatter,
+  formatter,
+  height,
+  showControls,
+  timelinePosition,
+  onTimelineUpdate,
+  timeframe,
+  onTimeframeUpdate,
+  isPlaying,
+  onPlay,
+  isPaused,
+  onPause,
+  onStop
+}) {
+  const [animationStep, setAnimationStep] = useState(0);
+
+  // Calculate animation step by duration
+  useEffect(() => {
+    if (duration && data.length) {
+      setAnimationStep(duration / data.length);
+    }
+  }, [duration, data, setAnimationStep]);
+
+  return (
```

Comment:
I'm curious why you are using context. Normally, context is used when data needs to be accessible by many components at different nesting levels. Is not your case, right?

Reply:
Good question. We need to use context inside TimeSeriesWidgetUI and useTimeSeriesInteractivity. If we don't use a context, we should have to bring every prop through TimeSeriesChart, something that I don't like. But it isn't my unique motivation for creating a context. I like TimeSeriesContext because there is where all the stuff about managing the widget state and it makes that logic transparent to whoever want to use it. Also, because it's concentrate in one place, an usere can create a context that replaces that one and might extends it.

The first approach didn't have that context, but when I saw all that logic inside the TimeSeriesWidgetUI and all that props inside the TimeSeriesChart... it was totally a mess. I preferred the context way, much cleaner.